In [1]:
import pandas as pd
from collections import Counter
import re

In [2]:
data1=pd.read_csv("../data/raw/rating_data.csv")

In [3]:
data2=pd.read_csv("../data/raw/reviews_data1.csv")

In [4]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3427 entries, 0 to 3426
Data columns (total 13 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Aircraft                1790 non-null   object 
 1   Type Of Traveller       2657 non-null   object 
 2   Seat Type               3425 non-null   object 
 3   Route                   2653 non-null   object 
 4   Date Flown              2649 non-null   object 
 5   Seat Comfort            3329 non-null   float64
 6   Cabin Staff Service     3322 non-null   float64
 7   Food & Beverages        3097 non-null   float64
 8   Inflight Entertainment  2415 non-null   float64
 9   Ground Service          2591 non-null   float64
 10  Wifi & Connectivity     506 non-null    float64
 11  Value For Money         3427 non-null   int64  
 12  Recommended             3427 non-null   object 
dtypes: float64(6), int64(1), object(6)
memory usage: 348.2+ KB


In [5]:
data1.isna().sum()

Aircraft                  1637
Type Of Traveller          770
Seat Type                    2
Route                      774
Date Flown                 778
Seat Comfort                98
Cabin Staff Service        105
Food & Beverages           330
Inflight Entertainment    1012
Ground Service             836
Wifi & Connectivity       2921
Value For Money              0
Recommended                  0
dtype: int64

In [6]:
data1.duplicated().sum()

np.int64(218)

In [7]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3427 entries, 0 to 3426
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Verified     3427 non-null   object
 1   Reviews      3427 non-null   object
 2   Recommended  3427 non-null   object
dtypes: object(3)
memory usage: 80.4+ KB


In [8]:
data2.isna().sum()

Verified       0
Reviews        0
Recommended    0
dtype: int64

In [9]:
data2.duplicated().sum()

np.int64(7)

In [10]:
data1.head()

,Aircraft,Type Of Traveller,Seat Type,Route,Date Flown,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Wifi & Connectivity,Value For Money,Recommended
0,Boeing 787-9,Business,Economy Class,Denver to London,December 2022,1.0,1.0,1.0,3.0,1.0,NaN,1,no
1,A320,Solo Leisure,Business Class,London to Marrakech,June 2022,3.0,5.0,5.0,NaN,4.0,NaN,3,yes
2,A380,Solo Leisure,Economy Class,San Francisco to London,November 2022,2.0,1.0,2.0,2.0,3.0,1.0,2,no
3,A320,Solo Leisure,Business Class,London to Lisbon,November 2022,3.0,4.0,4.0,NaN,3.0,NaN,3,yes
4,Boeing 787 / A320,Solo Leisure,Economy Class,Montreal to Edinburgh via London Heathrow,January 2022,4.0,4.0,4.0,4.0,4.0,NaN,4,yes


In [11]:
data2.head()

,Verified,Reviews,Recommended
0,yes,i was flying to warsaw for one day of mee...,no
1,yes,"booked a ba holiday to marrakech, after p...",yes
2,yes,extremely sub-par service. highlights: no ...,no
3,yes,i virtually gave up on british airways ab...,yes
4,yes,i was pleasantly surprised that the airli...,yes


# عدد الصفوف متساوي
# rating_data فيها 218 تكرار 
# reviews_data1 فيها 7 تكرار 
# عدد القيم الناقصه في reviews_data1 يساوي 0
# عدد القيم الناقصه في rating_data متنوع لكن اكثر الاعمده Wifi & Connectivity و Inflight Entertainment و Aircraft

In [12]:
len(data1['Recommended'])==len(data2['Recommended'])

True

In [13]:
data1['Recommended'].equals(data2['Recommended'])

True

تأكدنا ان صف Recommended في الملفين متساوي في الصفوف و الترتيب , الخطوه القادمه ندمج الملفين

In [14]:
data1.head()

,Aircraft,Type Of Traveller,Seat Type,Route,Date Flown,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Wifi & Connectivity,Value For Money,Recommended
0,Boeing 787-9,Business,Economy Class,Denver to London,December 2022,1.0,1.0,1.0,3.0,1.0,NaN,1,no
1,A320,Solo Leisure,Business Class,London to Marrakech,June 2022,3.0,5.0,5.0,NaN,4.0,NaN,3,yes
2,A380,Solo Leisure,Economy Class,San Francisco to London,November 2022,2.0,1.0,2.0,2.0,3.0,1.0,2,no
3,A320,Solo Leisure,Business Class,London to Lisbon,November 2022,3.0,4.0,4.0,NaN,3.0,NaN,3,yes
4,Boeing 787 / A320,Solo Leisure,Economy Class,Montreal to Edinburgh via London Heathrow,January 2022,4.0,4.0,4.0,4.0,4.0,NaN,4,yes


In [15]:
data2.head()

,Verified,Reviews,Recommended
0,yes,i was flying to warsaw for one day of mee...,no
1,yes,"booked a ba holiday to marrakech, after p...",yes
2,yes,extremely sub-par service. highlights: no ...,no
3,yes,i virtually gave up on british airways ab...,yes
4,yes,i was pleasantly surprised that the airli...,yes


In [16]:
data_clean_2=data2.drop(columns='Recommended')

In [17]:
merged_data = pd.concat(
    [
        data1.reset_index(drop=True),
        data_clean_2.reset_index(drop=True)
    ],
    axis=1
)

In [18]:
merged_data.to_csv("../data/raw/merged_raw.csv", index=False)

In [19]:
df=pd.read_csv('../data/raw/merged_raw.csv')

In [20]:
df.shape

(3427, 15)

In [21]:
df.columns

Index(['Aircraft', 'Type Of Traveller', 'Seat Type', 'Route', 'Date Flown',
       'Seat Comfort', 'Cabin Staff Service', 'Food & Beverages',
       'Inflight Entertainment', 'Ground Service', 'Wifi & Connectivity',
       'Value For Money', 'Recommended', 'Verified', 'Reviews'],
      dtype='object')

In [22]:
df.head()

,Aircraft,Type Of Traveller,Seat Type,Route,Date Flown,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Wifi & Connectivity,Value For Money,Recommended,Verified,Reviews
0,Boeing 787-9,Business,Economy Class,Denver to London,December 2022,1.0,1.0,1.0,3.0,1.0,NaN,1,no,yes,i was flying to warsaw for one day of mee...
1,A320,Solo Leisure,Business Class,London to Marrakech,June 2022,3.0,5.0,5.0,NaN,4.0,NaN,3,yes,yes,"booked a ba holiday to marrakech, after p..."
2,A380,Solo Leisure,Economy Class,San Francisco to London,November 2022,2.0,1.0,2.0,2.0,3.0,1.0,2,no,yes,extremely sub-par service. highlights: no ...
3,A320,Solo Leisure,Business Class,London to Lisbon,November 2022,3.0,4.0,4.0,NaN,3.0,NaN,3,yes,yes,i virtually gave up on british airways ab...
4,Boeing 787 / A320,Solo Leisure,Economy Class,Montreal to Edinburgh via London Heathrow,January 2022,4.0,4.0,4.0,4.0,4.0,NaN,4,yes,yes,i was pleasantly surprised that the airli...


In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3427 entries, 0 to 3426
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Aircraft                1790 non-null   object 
 1   Type Of Traveller       2657 non-null   object 
 2   Seat Type               3425 non-null   object 
 3   Route                   2653 non-null   object 
 4   Date Flown              2649 non-null   object 
 5   Seat Comfort            3329 non-null   float64
 6   Cabin Staff Service     3322 non-null   float64
 7   Food & Beverages        3097 non-null   float64
 8   Inflight Entertainment  2415 non-null   float64
 9   Ground Service          2591 non-null   float64
 10  Wifi & Connectivity     506 non-null    float64
 11  Value For Money         3427 non-null   int64  
 12  Recommended             3427 non-null   object 
 13  Verified                3427 non-null   object 
 14  Reviews                 3427 non-null   

In [24]:
df.isna().sum()

Aircraft                  1637
Type Of Traveller          770
Seat Type                    2
Route                      774
Date Flown                 778
Seat Comfort                98
Cabin Staff Service        105
Food & Beverages           330
Inflight Entertainment    1012
Ground Service             836
Wifi & Connectivity       2921
Value For Money              0
Recommended                  0
Verified                     0
Reviews                      0
dtype: int64

In [25]:
df.isna().mean() * 100

Aircraft                  47.767727
Type Of Traveller         22.468631
Seat Type                  0.058360
Route                     22.585352
Date Flown                22.702072
Seat Comfort               2.859644
Cabin Staff Service        3.063904
Food & Beverages           9.629413
Inflight Entertainment    29.530201
Ground Service            24.394514
Wifi & Connectivity       85.234899
Value For Money            0.000000
Recommended                0.000000
Verified                   0.000000
Reviews                    0.000000
dtype: float64

In [26]:
df.duplicated().sum()

np.int64(2)

In [27]:
df.nunique()

Aircraft                   191
Type Of Traveller            4
Seat Type                    4
Route                     1472
Date Flown                 101
Seat Comfort                 5
Cabin Staff Service          5
Food & Beverages             6
Inflight Entertainment       6
Ground Service               5
Wifi & Connectivity          5
Value For Money              6
Recommended                  2
Verified                     2
Reviews                   3409
dtype: int64

In [28]:
df["Recommended"].value_counts()

Recommended
no     1990
yes    1437
Name: count, dtype: int64

In [29]:
df["Verified"].value_counts()

Verified
no     2478
yes     949
Name: count, dtype: int64

In [30]:
df.describe()

,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Wifi & Connectivity,Value For Money
count,3329.000000,3322.000000,3097.000000,2415.000000,2591.000000,506.000000,3427.000000
mean,2.904776,3.278748,2.741040,2.643892,2.845619,1.934783,2.742048
std,1.361694,1.486654,1.441619,1.407423,1.443699,1.356209,1.465897
min,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000,0.000000
25%,2.000000,2.000000,1.000000,1.000000,1.000000,1.000000,1.000000
50%,3.000000,4.000000,3.000000,3.000000,3.000000,1.000000,3.000000
75%,4.000000,5.000000,4.000000,4.000000,4.000000,3.000000,4.000000
max,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000


نفحص القيم الي في التقييم تساوي 0 المنطق من 1 الى 5

In [31]:
(df['Food & Beverages'] == 0).sum()

np.int64(1)

In [32]:
(df['Inflight Entertainment']==0).sum()

np.int64(10)

In [33]:
(df['Value For Money']==0).sum()

np.int64(1)

In [34]:
clean_df =df.copy()

حذف الاعمده الي فيها نقص كبير 

In [35]:
clean_df=clean_df.drop(columns=["Aircraft", "Wifi & Connectivity"])

تحويل الاعمده الي فيها 0 الى nan

In [36]:
rating_zero_cols = [
    'Food & Beverages',
    'Inflight Entertainment',
    'Value For Money'
]

In [37]:
clean_df[rating_zero_cols] = clean_df[rating_zero_cols].replace(0, pd.NA)

In [38]:
(clean_df[rating_zero_cols] == 0).sum()

Food & Beverages          0
Inflight Entertainment    0
Value For Money           0
dtype: int64

In [39]:
clean_df[rating_zero_cols].isna().sum()

Food & Beverages           331
Inflight Entertainment    1022
Value For Money              1
dtype: int64

In [40]:
clean_df=clean_df.drop_duplicates()

In [41]:
clean_df.duplicated().sum()

np.int64(0)

In [42]:
clean_df.shape

(3425, 13)

In [43]:
clean_df.isna().sum()

Type Of Traveller          770
Seat Type                    2
Route                      774
Date Flown                 778
Seat Comfort                98
Cabin Staff Service        105
Food & Beverages           331
Inflight Entertainment    1022
Ground Service             836
Value For Money              1
Recommended                  0
Verified                     0
Reviews                      0
dtype: int64

In [44]:
clean_df.duplicated().sum()

np.int64(0)

In [45]:
clean_df.describe()

,Seat Comfort,Cabin Staff Service,Ground Service
count,3327.000000,3320.000000,2589.00000
mean,2.904118,3.278614,2.84550
std,1.361839,1.487042,1.44425
min,1.000000,1.000000,1.00000
25%,2.000000,2.000000,1.00000
50%,3.000000,4.000000,3.00000
75%,4.000000,5.000000,4.00000
max,5.000000,5.000000,5.00000


In [46]:
clean_df.dtypes

Type Of Traveller          object
Seat Type                  object
Route                      object
Date Flown                 object
Seat Comfort              float64
Cabin Staff Service       float64
Food & Beverages           object
Inflight Entertainment     object
Ground Service            float64
Value For Money            object
Recommended                object
Verified                   object
Reviews                    object
dtype: object

نرجع القيم رقميه للاعمده الي تحولت object بعد التعويض ب nan

In [47]:
rating_cols_to_fix = [
    'Food & Beverages',
    'Inflight Entertainment',
    'Value For Money'
]

In [48]:
clean_df[rating_cols_to_fix] = clean_df[rating_cols_to_fix].apply(pd.to_numeric)

In [49]:
clean_df.dtypes

Type Of Traveller          object
Seat Type                  object
Route                      object
Date Flown                 object
Seat Comfort              float64
Cabin Staff Service       float64
Food & Beverages          float64
Inflight Entertainment    float64
Ground Service            float64
Value For Money           float64
Recommended                object
Verified                   object
Reviews                    object
dtype: object

In [50]:
clean_df.describe()

,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Value For Money
count,3327.000000,3320.000000,3094.000000,2403.000000,2589.00000,3424.000000
mean,2.904118,3.278614,2.741758,2.655431,2.84550,2.742407
std,1.361839,1.487042,1.441461,1.400413,1.44425,1.465625
min,1.000000,1.000000,1.000000,1.000000,1.00000,1.000000
25%,2.000000,2.000000,1.000000,1.000000,1.00000,1.000000
50%,3.000000,4.000000,3.000000,3.000000,3.00000,3.000000
75%,4.000000,5.000000,4.000000,4.000000,4.00000,4.000000
max,5.000000,5.000000,5.000000,5.000000,5.00000,5.000000


In [51]:
pd.set_option("display.max_colwidth", None)

In [52]:
clean_df["Reviews"].head()

0         i was flying to warsaw for one day of meetings and would be flying back out to miami the next day. initially my flight was delayed in denver by 3 hours. as such i knew i would miss my connection in warsaw. the denver to heathrow flight was terrible. the plane was dirty, i was in economy, i usually fly business or premium economy but on this flight there was no availability. the plane was only about 2/3 full and the service was not good. my food was inedible and the seats very cramped although the middle seat was open so that made it tolerable. when i arrived in heathrow it was a complete circus. passengers everywhere trying to rebook missed connections. not only my flight but multiple other flights had missed connections. after waiting in line for an hour i was finally rebooked on another ba flight to warsaw. this plane was the smallest 3x3 plane i have seen. packed in like sardines. my bag never showed up so i had to cancel my meeting and now have no bag to fly back to miami

In [53]:
clean_df["Clean_Review"] = (
    clean_df["Reviews"]
    .fillna("")
    .str.strip()
    .str.replace(r"\s+", " ", regex=True)
)

In [54]:
clean_df["word_count"] = clean_df["Clean_Review"].str.split().str.len()

In [55]:
clean_df["Review_Length"] = clean_df["Reviews"].str.len()

In [56]:
pd.reset_option("display.max_colwidth")

In [57]:
clean_df

,Type Of Traveller,Seat Type,Route,Date Flown,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Value For Money,Recommended,Verified,Reviews,Clean_Review,word_count,Review_Length
0,Business,Economy Class,Denver to London,December 2022,1.0,1.0,1.0,3.0,1.0,1.0,no,yes,i was flying to warsaw for one day of mee...,i was flying to warsaw for one day of meetings...,261,1373
1,Solo Leisure,Business Class,London to Marrakech,June 2022,3.0,5.0,5.0,NaN,4.0,3.0,yes,yes,"booked a ba holiday to marrakech, after p...","booked a ba holiday to marrakech, after postin...",87,473
2,Solo Leisure,Economy Class,San Francisco to London,November 2022,2.0,1.0,2.0,2.0,3.0,2.0,no,yes,extremely sub-par service. highlights: no ...,extremely sub-par service. highlights: no onli...,141,784
3,Solo Leisure,Business Class,London to Lisbon,November 2022,3.0,4.0,4.0,NaN,3.0,3.0,yes,yes,i virtually gave up on british airways ab...,i virtually gave up on british airways about t...,78,459
4,Solo Leisure,Economy Class,Montreal to Edinburgh via London Heathrow,January 2022,4.0,4.0,4.0,4.0,4.0,4.0,yes,yes,i was pleasantly surprised that the airli...,i was pleasantly surprised that the airline co...,113,641
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3422,NaN,Premium Economy,NaN,NaN,4.0,3.0,3.0,4.0,NaN,4.0,yes,no,flight from heathrow to toronto. booked emerge...,flight from heathrow to toronto. booked emerge...,68,371
3423,NaN,Business Class,NaN,NaN,4.0,5.0,4.0,NaN,NaN,3.0,yes,no,lhr to ham. purser addresses all club passenge...,lhr to ham. purser addresses all club passenge...,25,144
3424,NaN,Economy Class,NaN,NaN,NaN,NaN,NaN,NaN,NaN,4.0,yes,no,my son who had worked for british airways urge...,my son who had worked for british airways urge...,230,1171
3425,NaN,Premium Economy,NaN,NaN,1.0,3.0,5.0,NaN,NaN,1.0,no,no,london city-new york jfk via shannon on a318 b...,london city-new york jfk via shannon on a318 b...,33,188


In [58]:
clean_df.describe()

,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Value For Money,word_count,Review_Length
count,3327.000000,3320.000000,3094.000000,2403.000000,2589.00000,3424.000000,3425.000000,3425.000000
mean,2.904118,3.278614,2.741758,2.655431,2.84550,2.742407,160.888759,888.892847
std,1.361839,1.487042,1.441461,1.400413,1.44425,1.465625,105.119062,568.223378
min,1.000000,1.000000,1.000000,1.000000,1.00000,1.000000,15.000000,83.000000
25%,2.000000,2.000000,1.000000,1.000000,1.00000,1.000000,86.000000,485.000000
50%,3.000000,4.000000,3.000000,3.000000,3.00000,3.000000,134.000000,747.000000
75%,4.000000,5.000000,4.000000,4.000000,4.00000,4.000000,203.000000,1120.000000
max,5.000000,5.000000,5.000000,5.000000,5.00000,5.000000,655.000000,3537.000000


In [59]:
pd.set_option("display.max_colwidth", None)

In [60]:
clean_df[["Reviews", "Clean_Review"]].head()

,Reviews,Clean_Review
0,"i was flying to warsaw for one day of meetings and would be flying back out to miami the next day. initially my flight was delayed in denver by 3 hours. as such i knew i would miss my connection in warsaw. the denver to heathrow flight was terrible. the plane was dirty, i was in economy, i usually fly business or premium economy but on this flight there was no availability. the plane was only about 2/3 full and the service was not good. my food was inedible and the seats very cramped although the middle seat was open so that made it tolerable. when i arrived in heathrow it was a complete circus. passengers everywhere trying to rebook missed connections. not only my flight but multiple other flights had missed connections. after waiting in line for an hour i was finally rebooked on another ba flight to warsaw. this plane was the smallest 3x3 plane i have seen. packed in like sardines. my bag never showed up so i had to cancel my meeting and now have no bag to fly back to miami. ba has shown no desire to help other than to give me standard answers and that hopefully my bag will show up in miami at some point and i can retrieve it from there. i mean the bag is sitting in heathrow somewhere with 6 different fight options to get it to warsaw yet ba in unwilling to do anything to expedite getting my bag to warsaw before i leave for miami the next day.","i was flying to warsaw for one day of meetings and would be flying back out to miami the next day. initially my flight was delayed in denver by 3 hours. as such i knew i would miss my connection in warsaw. the denver to heathrow flight was terrible. the plane was dirty, i was in economy, i usually fly business or premium economy but on this flight there was no availability. the plane was only about 2/3 full and the service was not good. my food was inedible and the seats very cramped although the middle seat was open so that made it tolerable. when i arrived in heathrow it was a complete circus. passengers everywhere trying to rebook missed connections. not only my flight but multiple other flights had missed connections. after waiting in line for an hour i was finally rebooked on another ba flight to warsaw. this plane was the smallest 3x3 plane i have seen. packed in like sardines. my bag never showed up so i had to cancel my meeting and now have no bag to fly back to miami. ba has shown no desire to help other than to give me standard answers and that hopefully my bag will show up in miami at some point and i can retrieve it from there. i mean the bag is sitting in heathrow somewhere with 6 different fight options to get it to warsaw yet ba in unwilling to do anything to expedite getting my bag to warsaw before i leave for miami the next day."
1,"booked a ba holiday to marrakech, after posting a negative review regarding my double tier points credit, i can now confirm that the points have been credited. it took a while but ba did eventually make good on their promise, thank you. based on things now being resolved i have travelled on ba since and i am booked to travel to south africa in first class on 2nd january, looking forward to the new ba concorde lounge. keep it up ba, loyalty is indeed a 2way street.","booked a ba holiday to marrakech, after posting a negative review regarding my double tier points credit, i can now confirm that the points have been credited. it took a while but ba did eventually make good on their promise, thank you. based on things now being resolved i have travelled on ba since and i am booked to travel to south africa in first class on 2nd january, looking forward to the new ba concorde lounge. keep it up ba, loyalty is indeed a 2way street."
2,"extremely sub-par service. highlights: no online meal selection. staff pretty much dropped off whatever meal they had and were very rude if a passenger wanted it changed. last meal was delayed to the last 40 minutes of flight. plates were literally grabbed away. i had to hand over my unfi

In [61]:
(clean_df["Clean_Review"] == "").sum()

np.int64(0)

In [62]:
clean_df.sort_values("word_count")[["Clean_Review", "word_count", "Review_Length"]].head()

,Clean_Review,word_count,Review_Length
2502,no problems on this short flight. left and arrived on time. drink and snack served.,15,83
3394,rating : 10/10 cabin flown economy value for money seat comfort staff service catering entertainment recommended,16,112
2988,lhr-bgo on a319. flight on time and only half full. drinks and snack on this early departure. crew very good.,20,109
3070,gig-lhr b773 newly refurbished cabin. seats were good and comfortable food was reasonable and entertainment was great hd screen with many options.,22,146
429,gatwick to st lucia. great service on this flight. thank you very much for the cabin and the great cabin crew. excellent service.,23,134


نسوي عمود يحسب متوسط التقييمات

In [63]:
rating_cols = [
    "Seat Comfort",
    "Cabin Staff Service",
    "Food & Beverages",
    "Inflight Entertainment",
    "Ground Service",
    "Value For Money"
]

In [64]:
clean_df["Average_Rating"] = clean_df[rating_cols].mean(axis=1)

In [65]:
clean_df[[
    "Seat Comfort",
    "Cabin Staff Service",
    "Food & Beverages",
    "Inflight Entertainment",
    "Ground Service",
    "Value For Money",
    "Average_Rating"
]].head()

,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Value For Money,Average_Rating
0,1.0,1.0,1.0,3.0,1.0,1.0,1.333333
1,3.0,5.0,5.0,NaN,4.0,3.0,4.000000
2,2.0,1.0,2.0,2.0,3.0,2.0,2.000000
3,3.0,4.0,4.0,NaN,3.0,3.0,3.400000
4,4.0,4.0,4.0,4.0,4.0,4.0,4.000000


In [66]:
clean_df["Average_Rating"].describe()

count    3425.000000
mean        2.833718
std         1.196549
min         1.000000
25%         1.833333
50%         2.750000
75%         3.833333
max         5.000000
Name: Average_Rating, dtype: float64

نظيف عمود Recommendation_Status

In [67]:
clean_df["Recommendation_Status"] = clean_df["Recommended"].replace({
    "yes": "Recommended",
    "no": "Not Recommended"
})

In [68]:
clean_df[["Recommended", "Recommendation_Status"]].head()

,Recommended,Recommendation_Status
0,no,Not Recommended
1,yes,Recommended
2,no,Not Recommended
3,yes,Recommended
4,yes,Recommended


انشاء عمود Verification_Status

In [69]:
clean_df['Verification_Status']=clean_df['Verified'].replace({
    "yes":"Verified",
    "no":"Not Verified"
})

In [70]:
clean_df[["Verified", "Verification_Status"]].head()

,Verified,Verification_Status
0,yes,Verified
1,yes,Verified
2,yes,Verified
3,yes,Verified
4,yes,Verified


نقسم Rating_Level بناءً على عمود Average_Rating

In [71]:
clean_df["Rating_Level"] = pd.cut(
    clean_df["Average_Rating"],
    bins=[0, 3, 4, 5.1],
    labels=["Low Rating", "Medium Rating", "High Rating"],
    include_lowest=True,
    right=False
)

In [72]:
clean_df["Rating_Level"].value_counts()

Rating_Level
Low Rating       1875
High Rating       839
Medium Rating     711
Name: count, dtype: int64

In [73]:
clean_df["Rating_Level"].isna().sum()

np.int64(0)

In [74]:
clean_df[["Average_Rating", "Rating_Level"]].head(10)

,Average_Rating,Rating_Level
0,1.333333,Low Rating
1,4.000000,High Rating
2,2.000000,Low Rating
3,3.400000,Medium Rating
4,4.000000,High Rating
5,2.333333,Low Rating
6,2.400000,Low Rating
7,1.000000,Low Rating
8,2.166667,Low Rating
9,2.333333,Low Rating


In [75]:
clean_df[['Average_Rating', 'Recommendation_Status', 'Verification_Status', 'Rating_Level']].head()

,Average_Rating,Recommendation_Status,Verification_Status,Rating_Level
0,1.333333,Not Recommended,Verified,Low Rating
1,4.000000,Recommended,Verified,High Rating
2,2.000000,Not Recommended,Verified,Low Rating
3,3.400000,Recommended,Verified,Medium Rating
4,4.000000,Recommended,Verified,High Rating


In [76]:
clean_df['Average_Rating'].describe()

count    3425.000000
mean        2.833718
std         1.196549
min         1.000000
25%         1.833333
50%         2.750000
75%         3.833333
max         5.000000
Name: Average_Rating, dtype: float64

In [77]:
clean_df["Rating_Level"].value_counts()

Rating_Level
Low Rating       1875
High Rating       839
Medium Rating     711
Name: count, dtype: int64

In [78]:
clean_df['Recommendation_Status'].value_counts()

Recommendation_Status
Not Recommended    1990
Recommended        1435
Name: count, dtype: int64

In [79]:
clean_df['Verification_Status'].value_counts()

Verification_Status
Not Verified    2476
Verified         949
Name: count, dtype: int64

نفحص عمود Date Flown

In [80]:
clean_df["Date Flown"].head(10)

0    December 2022
1        June 2022
2    November 2022
3    November 2022
4     January 2022
5    November 2022
6    November 2022
7      August 2022
8    November 2022
9    November 2022
Name: Date Flown, dtype: object

In [81]:
clean_df["Date Flown"].value_counts().head(20)

Date Flown
August 2015       83
September 2015    78
April 2017        65
November 2015     64
October 2015      64
January 2017      60
November 2016     58
October 2016      57
December 2015     57
September 2016    56
June 2017         56
July 2015         56
December 2016     55
February 2017     52
March 2016        51
May 2016          50
March 2017        49
August 2016       47
May 2017          47
September 2017    44
Name: count, dtype: int64

In [82]:
clean_df["Date_Flown_Clean"] = pd.to_datetime(
    clean_df["Date Flown"],
    format="%B %Y",
    errors="coerce"
)

In [83]:
clean_df[["Date Flown", "Date_Flown_Clean"]].head(10)

,Date Flown,Date_Flown_Clean
0,December 2022,2022-12-01
1,June 2022,2022-06-01
2,November 2022,2022-11-01
3,November 2022,2022-11-01
4,January 2022,2022-01-01
5,November 2022,2022-11-01
6,November 2022,2022-11-01
7,August 2022,2022-08-01
8,November 2022,2022-11-01
9,November 2022,2022-11-01


In [84]:
clean_df["Date_Flown_Clean"].isna().sum()

np.int64(778)

In [85]:
clean_df["Year"] = clean_df["Date_Flown_Clean"].dt.year

In [86]:
clean_df["Month"] = clean_df["Date_Flown_Clean"].dt.month

In [87]:
clean_df["Month_Name"] = clean_df["Date_Flown_Clean"].dt.month_name()

In [88]:
clean_df[["Date Flown", "Date_Flown_Clean", "Year", "Month", "Month_Name"]].head(10)

,Date Flown,Date_Flown_Clean,Year,Month,Month_Name
0,December 2022,2022-12-01,2022.0,12.0,December
1,June 2022,2022-06-01,2022.0,6.0,June
2,November 2022,2022-11-01,2022.0,11.0,November
3,November 2022,2022-11-01,2022.0,11.0,November
4,January 2022,2022-01-01,2022.0,1.0,January
5,November 2022,2022-11-01,2022.0,11.0,November
6,November 2022,2022-11-01,2022.0,11.0,November
7,August 2022,2022-08-01,2022.0,8.0,August
8,November 2022,2022-11-01,2022.0,11.0,November
9,November 2022,2022-11-01,2022.0,11.0,November


In [89]:
clean_df["Year"].value_counts().sort_index()

Year
2014.0      9
2015.0    472
2016.0    563
2017.0    560
2018.0    338
2019.0    336
2020.0     94
2021.0     93
2022.0    182
Name: count, dtype: int64

In [90]:
clean_df["Year"] = clean_df["Year"].astype("Int64")

In [91]:
clean_df["Month"] = clean_df["Month"].astype("Int64")

In [92]:
clean_df[["Date Flown", "Date_Flown_Clean", "Year", "Month", "Month_Name"]].head(10)

,Date Flown,Date_Flown_Clean,Year,Month,Month_Name
0,December 2022,2022-12-01,2022,12,December
1,June 2022,2022-06-01,2022,6,June
2,November 2022,2022-11-01,2022,11,November
3,November 2022,2022-11-01,2022,11,November
4,January 2022,2022-01-01,2022,1,January
5,November 2022,2022-11-01,2022,11,November
6,November 2022,2022-11-01,2022,11,November
7,August 2022,2022-08-01,2022,8,August
8,November 2022,2022-11-01,2022,11,November
9,November 2022,2022-11-01,2022,11,November


فحص بعض الاعمده 
Type Of Traveller - Seat Type - Route

In [93]:
clean_df["Type Of Traveller"].value_counts(dropna=False)

Type Of Traveller
Couple Leisure    894
Solo Leisure      812
NaN               770
Business          595
Family Leisure    354
Name: count, dtype: int64

In [94]:
clean_df["Seat Type"].value_counts(dropna=False)

Seat Type
Economy Class      1767
Business Class     1113
Premium Economy     337
First Class         206
NaN                   2
Name: count, dtype: int64

In [95]:
clean_df["Route"].value_counts(dropna=False).head(20)

Route
NaN                       774
London to Johannesburg     16
London to Cape Town        15
Vancouver to London        13
BKK to LHR                 13
LHR to MIA                 12
Johannesburg to London     12
London to Athens           11
Cape Town to London        11
LHR to BKK                 10
JFK to LHR                 10
Hong Kong to London        10
LHR to SFO                  9
Istanbul to London          9
LHR to HKG                  9
LHR to CPT                  9
HKG to LHR                  9
London to Hong Kong         9
London to Toronto           9
Toronto to London           9
Name: count, dtype: int64

In [96]:
clean_df['Type Of Traveller']=clean_df['Type Of Traveller'].fillna("Unknown")

In [97]:
clean_df["Seat Type"] = clean_df["Seat Type"].fillna("Unknown")

In [98]:
clean_df["Route"] = clean_df["Route"].fillna("Unknown")

In [99]:
clean_df['Type Of Traveller'].isna().sum()

np.int64(0)

In [100]:
clean_df['Seat Type'].isna().sum()

np.int64(0)

In [101]:
clean_df['Route'].isna().sum()

np.int64(0)

In [102]:
clean_df["Type Of Traveller"].value_counts(dropna=False)

Type Of Traveller
Couple Leisure    894
Solo Leisure      812
Unknown           770
Business          595
Family Leisure    354
Name: count, dtype: int64

In [103]:
clean_df["Seat Type"].value_counts(dropna=False)

Seat Type
Economy Class      1767
Business Class     1113
Premium Economy     337
First Class         206
Unknown               2
Name: count, dtype: int64

In [104]:
clean_df["Route"].value_counts(dropna=False).head(20)

Route
Unknown                   774
London to Johannesburg     16
London to Cape Town        15
Vancouver to London        13
BKK to LHR                 13
LHR to MIA                 12
Johannesburg to London     12
London to Athens           11
Cape Town to London        11
LHR to BKK                 10
JFK to LHR                 10
Hong Kong to London        10
LHR to SFO                  9
Istanbul to London          9
LHR to HKG                  9
LHR to CPT                  9
HKG to LHR                  9
London to Hong Kong         9
London to Toronto           9
Toronto to London           9
Name: count, dtype: int64

فحص القيم الناقصة المتبقية في الداتا كلها بعد كل التنظيف

In [105]:
clean_df.isna().sum()

Type Of Traveller            0
Seat Type                    0
Route                        0
Date Flown                 778
Seat Comfort                98
Cabin Staff Service        105
Food & Beverages           331
Inflight Entertainment    1022
Ground Service             836
Value For Money              1
Recommended                  0
Verified                     0
Reviews                      0
Clean_Review                 0
word_count                   0
Review_Length                0
Average_Rating               0
Recommendation_Status        0
Verification_Status          0
Rating_Level                 0
Date_Flown_Clean           778
Year                       778
Month                      778
Month_Name                 778
dtype: int64

In [106]:
clean_df.shape

(3425, 24)

In [107]:
clean_df.dtypes

Type Of Traveller                 object
Seat Type                         object
Route                             object
Date Flown                        object
Seat Comfort                     float64
Cabin Staff Service              float64
Food & Beverages                 float64
Inflight Entertainment           float64
Ground Service                   float64
Value For Money                  float64
Recommended                       object
Verified                          object
Reviews                           object
Clean_Review                      object
word_count                         int64
Review_Length                      int64
Average_Rating                   float64
Recommendation_Status             object
Verification_Status               object
Rating_Level                    category
Date_Flown_Clean          datetime64[ns]
Year                               Int64
Month                              Int64
Month_Name                        object
dtype: object

In [108]:
clean_df = clean_df.rename(columns={
    "word_count": "Word_Count"
})

In [109]:
clean_df

,Type Of Traveller,Seat Type,Route,Date Flown,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Value For Money,...,Word_Count,Review_Length,Average_Rating,Recommendation_Status,Verification_Status,Rating_Level,Date_Flown_Clean,Year,Month,Month_Name
0,Business,Economy Class,Denver to London,December 2022,1.0,1.0,1.0,3.0,1.0,1.0,...,261,1373,1.333333,Not Recommended,Verified,Low Rating,2022-12-01,2022,12,December
1,Solo Leisure,Business Class,London to Marrakech,June 2022,3.0,5.0,5.0,NaN,4.0,3.0,...,87,473,4.000000,Recommended,Verified,High Rating,2022-06-01,2022,6,June
2,Solo Leisure,Economy Class,San Francisco to London,November 2022,2.0,1.0,2.0,2.0,3.0,2.0,...,141,784,2.000000,Not Recommended,Verified,Low Rating,2022-11-01,2022,11,November
3,Solo Leisure,Business Class,London to Lisbon,November 2022,3.0,4.0,4.0,NaN,3.0,3.0,...,78,459,3.400000,Recommended,Verified,Medium Rating,2022-11-01,2022,11,November
4,Solo Leisure,Economy Class,Montreal to Edinburgh via London Heathrow,January 2022,4.0,4.0,4.0,4.0,4.0,4.0,...,113,641,4.000000,Recommended,Verified,High Rating,2022-01-01,2022,1,January
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3422,Unknown,Premium Economy,Unknown,NaN,4.0,3.0,3.0,4.0,NaN,4.0,...,68,371,3.600000,Recommended,Not Verified,Medium Rating,NaT,<NA>,<NA>,NaN
3423,Unknown,Business Class,Unknown,NaN,4.0,5.0,4.0,NaN,NaN,3.0,...,25,144,4.000000,Recommended,Not Verified,High Rating,NaT,<NA>,<NA>,NaN
3424,Unknown,Economy Class,Unknown,NaN,NaN,NaN,NaN,NaN,NaN,4.0,...,230,1171,4.000000,Recommended,Not Verified,High Rating,NaT,<NA>,<NA>,NaN
3425,Unknown,Premium Economy,Unknown,NaN,1.0,3.0,5.0,NaN,NaN,1.0,...,33,188,2.500000,Not Recommended,Not Verified,Low Rating,NaT,<NA>,<NA>,NaN


In [110]:
clean_df[["Word_Count", "Review_Length"]].head()

,Word_Count,Review_Length
0,261,1373
1,87,473
2,141,784
3,78,459
4,113,641


In [111]:
clean_df.to_csv("../data/processed/ba_reviews_cleaned.csv", index=False)

PermissionError: [Errno 13] Permission denied: '../data/processed/ba_reviews_cleaned.csv'

In [ ]:
test_df = pd.read_csv("../data/processed/ba_reviews_cleaned.csv")

In [ ]:
test_df.shape

(3425, 24)

In [ ]:
test_df.head()

,Type Of Traveller,Seat Type,Route,Date Flown,Seat Comfort,Cabin Staff Service,Food & Beverages,Inflight Entertainment,Ground Service,Value For Money,...,Word_Count,Review_Length,Average_Rating,Recommendation_Status,Verification_Status,Rating_Level,Date_Flown_Clean,Year,Month,Month_Name
0,Business,Economy Class,Denver to London,December 2022,1.0,1.0,1.0,3.0,1.0,1.0,...,261,1373,1.333333,Not Recommended,Verified,Low Rating,2022-12-01,2022.0,12.0,December
1,Solo Leisure,Business Class,London to Marrakech,June 2022,3.0,5.0,5.0,NaN,4.0,3.0,...,87,473,4.000000,Recommended,Verified,High Rating,2022-06-01,2022.0,6.0,June
2,Solo Leisure,Economy Class,San Francisco to London,November 2022,2.0,1.0,2.0,2.0,3.0,2.0,...,141,784,2.000000,Not Recommended,Verified,Low Rating,2022-11-01,2022.0,11.0,November
3,Solo Leisure,Business Class,London to Lisbon,November 2022,3.0,4.0,4.0,NaN,3.0,3.0,...,78,459,3.400000,Recommended,Verified,Medium Rating,2022-11-01,2022.0,11.0,November
4,Solo Leisure,Economy Class,Montreal to Edinburgh via London Heathrow,January 2022,4.0,4.0,4.0,4.0,4.0,4.0,...,113,641,4.000000,Recommended,Verified,High Rating,2022-01-01,2022.0,1.0,January


مرحلة EDA

In [ ]:
Total_Reviews=len(clean_df)
Total_Reviews

3425

In [ ]:
clean_df['Average_Rating'].mean()

np.float64(2.833717761557178)

In [ ]:
clean_df["Recommendation_Status"].value_counts()

Recommendation_Status
Not Recommended    1990
Recommended        1435
Name: count, dtype: int64

In [ ]:
clean_df['Recommendation_Status'].value_counts(normalize=True)*100

Recommendation_Status
Not Recommended    58.10219
Recommended        41.89781
Name: proportion, dtype: float64

In [ ]:
clean_df["Rating_Level"].value_counts()

Rating_Level
Low Rating       1875
High Rating       839
Medium Rating     711
Name: count, dtype: int64

In [ ]:
clean_df["Verification_Status"].value_counts()

Verification_Status
Not Verified    2476
Verified         949
Name: count, dtype: int64

In [ ]:
clean_df['Word_Count'].mean()

np.float64(160.88875912408758)

In [ ]:
clean_df['Review_Length'].mean()

np.float64(888.8928467153285)

التقييم حسب الفئات

In [ ]:
clean_df.groupby("Seat Type")["Average_Rating"].mean().sort_values(ascending=False)

Seat Type
First Class        3.386974
Business Class     2.965064
Premium Economy    2.861375
Economy Class      2.682381
Unknown            1.800000
Name: Average_Rating, dtype: float64

In [ ]:
clean_df.groupby("Type Of Traveller")["Average_Rating"].mean().sort_values(ascending=False)

Type Of Traveller
Unknown           3.152576
Solo Leisure      2.969520
Couple Leisure    2.787192
Family Leisure    2.742373
Business          2.360000
Name: Average_Rating, dtype: float64

In [ ]:
clean_df["Seat Type"].value_counts()

Seat Type
Economy Class      1767
Business Class     1113
Premium Economy     337
First Class         206
Unknown               2
Name: count, dtype: int64

In [ ]:
clean_df["Type Of Traveller"].value_counts()

Type Of Traveller
Couple Leisure    894
Solo Leisure      812
Unknown           770
Business          595
Family Leisure    354
Name: count, dtype: int64

In [ ]:
rating_cols = [
    "Seat Comfort",
    "Cabin Staff Service",
    "Food & Beverages",
    "Inflight Entertainment",
    "Ground Service",
    "Value For Money"
]

clean_df[rating_cols].mean().sort_values()

Inflight Entertainment    2.655431
Food & Beverages          2.741758
Value For Money           2.742407
Ground Service            2.845500
Seat Comfort              2.904118
Cabin Staff Service       3.278614
dtype: float64

العلاقة بين التوصية ومتوسط التقييم

In [ ]:
clean_df.groupby("Recommendation_Status")["Average_Rating"].mean()

Recommendation_Status
Not Recommended    2.007554
Recommended        3.979408
Name: Average_Rating, dtype: float64

In [ ]:
clean_df.groupby("Recommendation_Status")["Word_Count"].mean()

Recommendation_Status
Not Recommended    177.017085
Recommended        138.522648
Name: Word_Count, dtype: float64

الكلمات المتكررة في النصوص

In [ ]:
stop_words = {
    "the", "and", "to", "of", "a", "in", "was", "were", "is", "it", "for",
    "on", "with", "that", "this", "as", "at", "we", "i", "my", "our", "they",
    "you", "but", "had", "have", "has", "be", "been", "are", "not", "from",
    "by", "or", "an", "so", "if", "all", "very", "there", "their", "them",
    "he", "she", "his", "her", "me", "us", "will", "would", "could", "should",
    "can", "just", "do", "did", "does", "no", "yes"
}

custom_stop_words = {
    "flight", "flights", "british", "airways", "airway", "ba", "london",
    "which", "when", "one", "only", "get", "also", "even",
    "back", "way", "much", "many", "really",
    "out", "after", "first", "more", "than", "other", "heathrow",
    "class", "business", "economy", "cabin", "return", "check"
}

stop_words = stop_words.union(custom_stop_words)

all_words = []

for review in clean_df["Clean_Review"]:
    review = re.sub(r"[^a-zA-Z\s]", " ", review)
    words = review.lower().split()

    for word in words:
        if word not in stop_words and len(word) > 2:
            all_words.append(word)

word_counts = Counter(all_words)

top_words_df = pd.DataFrame(
    word_counts.most_common(20),
    columns=["Word", "Count"]
)

top_words_df

,Word,Count
0,service,2619
1,seat,2279
2,food,2249
3,crew,2054
4,seats,1982
5,good,1973
6,time,1942
7,staff,1457
8,club,1037
9,lhr,1018


In [ ]:
extra_stop_words = {
    "airline", "again", "about", "review", "then", "fly", "what",
    "verified", "told", "plane", "passengers", "new"
}

stop_words = stop_words.union(extra_stop_words)

In [ ]:
recommended_reviews = clean_df[
    clean_df["Recommendation_Status"] == "Recommended"
]["Clean_Review"]

In [ ]:
all_words_recommended = []

for review in recommended_reviews:
    review = re.sub(r"[^a-zA-Z\s]", " ", review)
    words = review.lower().split()

    for word in words:
        if word not in stop_words and len(word) > 2:
            all_words_recommended.append(word)

recommended_word_counts = Counter(all_words_recommended)

top_recommended_words = pd.DataFrame(
    recommended_word_counts.most_common(20),
    columns=["Word", "Count"]
)

top_recommended_words

,Word,Count
0,good,1298
1,crew,1072
2,service,1056
3,food,995
4,seat,943
5,time,923
6,seats,786
7,lhr,551
8,club,533
9,lounge,516


In [ ]:
not_recommended_reviews = clean_df[
    clean_df["Recommendation_Status"] == "Not Recommended"
]["Clean_Review"]

In [ ]:
all_words_not_recommended = []

for review in not_recommended_reviews:
    review = re.sub(r"[^a-zA-Z\s]", " ", review)
    words = review.lower().split()

    for word in words:
        if word not in stop_words and len(word) > 2:
            all_words_not_recommended.append(word)

not_recommended_word_counts = Counter(all_words_not_recommended)

top_not_recommended_words = pd.DataFrame(
    not_recommended_word_counts.most_common(20),
    columns=["Word", "Count"]
)

top_not_recommended_words

,Word,Count
0,service,1563
1,seat,1336
2,food,1254
3,seats,1196
4,time,1019
5,crew,982
6,staff,943
7,hours,715
8,good,675
9,before,540


In [115]:
import csv

In [116]:
powerbi_df = clean_df.drop(columns=["Reviews", "Clean_Review"])

In [117]:
powerbi_df.to_csv(
    "../data/processed/ba_reviews_powerbi_main.csv",
    index=False,
    encoding="utf-8-sig"
)

In [118]:
test_powerbi = pd.read_csv("../data/processed/ba_reviews_powerbi_main.csv")
test_powerbi.shape

(3425, 22)